stdlib/public/Darwin/Foundation/NSStringAPI.swift

```diff
@@ -381,6 +395,14 @@ extension String {
   /// Returns a `String` initialized by converting given `data` into
   /// Unicode characters using a given `encoding`.
   public init?(data: __shared Data, encoding: Encoding) {
+    if encoding == .utf8,
+       let str = data.withUnsafeBytes({
+         String._tryFromUTF8($0.bindMemory(to: UInt8.self))
```

Comment:
Mm, in previous discussions, we've agreed that the safety of binding `Data` is dubious given the fact that the underlying buffer may have been previously bound. Are we making the assertion that binding to a trivial type here is safe? /CC @atrick

Reply:
Is it possible to express the UTF-8 validation in terms of raw buffer pointers instead of typed? (Or overload to make this possible safely?)

Reply:
I don't know if this is something to be concerned with, but if it is:

The fix would be for Data to provide a way to access its contents as a `UBP<UInt8>`, since Data is the owner (or knows who the owner is) and is the only one allowed to do so. If String exposed something taking a `URBP`, then treated it as a collection of `UInt8`s, that seems like it's sweeping the potential time bomb under the living room rug.

Reply:
`Data` might know who the owner is, but it doesn't necessarily make it safe to bind the contents to `UInt8`, since the previous owner may have bound the buffer to a nontrivial type, and `Data` cannot unbind it — `URBP`s specifically do allow reading as `UInt8`s via a raw interface, though, making it always safe to read from.

I'm not _terribly_ concerned (as we've previously discussed — we can't prevent someone from doing something like reading arbitrary data bound to some non-trivial type `U` and trying to pass it along to be interpreted as UTF-8), but we did do a lot in `Data` to make sure we don't invoke undefined behavior like this silently on others' behalf. @atrick might have stronger opinions here, but I agree with him in that we shouldn't be repeating that here.

Reply:
FYI `withContiguousStorageIfAvailable` provides a typed pointer, so Data could use that. Unfortunately, `URBP` doesn't support it (I guess because it doesn't want to bind random memory).

This is also an issue with https://github.com/apple/swift/pull/22028 - currently, if you write:

```
class MyClass {}
var s = [MyClass(), MyClass(), MyClass()]
let sbytes = withUnsafeBytes(of: &s) { String(decoding: $0, as: UTF8.self) }
```
you will invoke a fast-path which calls `assumingMemoryBound(to: UInt8.self)` on the raw pointer, which IIUC is technically undefined behaviour. But it was marked "unsafe"...

Reply:
Any String utilities that we want to accept a raw byte stream, like Data, including validateUTF8 and _uncheckedFromUTF8 should either take URBP or some ContiguouslyStored<UInt8> protocol. Either will be source compatible with the existing APIs, so they can probably be fixed later.